src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestTools.UnitTesting.Attributes
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Reflection;
+
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    /// <summary>
+    /// Enum to specify whether the data is stored as property or in method.
+    /// </summary>
+    public enum DynamicDataSourceType
+    {
+        /// <summary>
+        /// Data is declared as property.
+        /// </summary>
+        Property = 0,
+
+        /// <summary>
+        /// Data is declared in method.
+        /// </summary>
+        Method = 1
+    }
+
+    /// <summary>
+    /// Attribute to define dynamic data for a test method.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+    public sealed class DynamicDataAttribute : TestDataSourceAttribute
+    {
+        private string dynamicDataSourceName;
+
+        private Type dynamicDataDeclaringType;
+
+        private DynamicDataSourceType dynamicDataSourceType;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DynamicDataAttribute"/> class.
+        /// </summary>
+        /// <param name="dynamicDataSourceName">
+        /// The name of method or property having test data.
+        /// </param>
+        /// <param name="dynamicDataSourceType">
+        /// Specifies whether the data is stored as property or in method.
+        /// </param>
+        public DynamicDataAttribute(string dynamicDataSourceName, DynamicDataSourceType dynamicDataSourceType = DynamicDataSourceType.Property)
+        {
+            this.dynamicDataSourceName = dynamicDataSourceName;
+            this.dynamicDataSourceType = dynamicDataSourceType;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DynamicDataAttribute"/> class.
+        /// </summary>
+        /// <param name="dynamicDataSourceName">
+        /// The type in which the test data is declared (as property or in method).
+        /// </param>
+        /// <param name="dynamicDataDeclaringType">
+        /// The declaring type of property or method having data.
+        /// </param>
+        /// <param name="dynamicDataSourceType">
+        /// Specifies whether the data is stored as property or in method.
+        /// </param>
+        public DynamicDataAttribute(string dynamicDataSourceName, Type dynamicDataDeclaringType, DynamicDataSourceType dynamicDataSourceType = DynamicDataSourceType.Property)
```

Comment:
Suggestion :  you can call above constructor to set sourceName and sourceType by using "this" keyword.